activemq_info.py

```diff
@@ -6,84 +7,131 @@
+        self.verbose_logging = verbose_logging
 
     def log_verbose(self, msg):
         if not self.verbose_logging:
             return
-        collectd.info('activemq_info plugin [verbose]: %s' % msg)
+        log.info('activemq_info plugin [verbose]: %s' % msg)
+
+    def fetch_metrics(self):
+        """Connect to ActiveMQ admin webpage and return DOM object"""
+        url = 'http://%s:%s/admin/xml/queues.jsp' % (self.host, self.port)
+        dom = None
+        try:
+            dom = minidom.parse(urllib.urlopen(url))
+        except Exception:
+            self.log_verbose('activemq_info plugin: No info received, '
+                             'offline node or turned off ActiveMQ')
+            return
+
+        queuenodes = dom.getElementsByTagName("queue")
+
+        gauges = []
+        counters = []
+
+        for node in queuenodes:
+            queue = node.attributes.item(0).value.replace('.', '_')
+            stats_item = node.getElementsByTagName('stats').item(0)
+            size = stats_item.getAttribute('size')
+            consumerCount = stats_item.getAttribute('consumerCount')
+            enqueueCount = stats_item.getAttribute('enqueueCount')
+            dequeueCount = stats_item.getAttribute('dequeueCount')
+
+            gauges.append((queue, 'size', size))
+            gauges.append((queue, 'consumerCount', consumerCount))
+            counters.append(queue, 'enqueueCount',  enqueueCount)
+            counters.append(queue, 'dequeueCount',  dequeueCount)
+
+        metrics = {
+            'gauges': gauges,
+            'counters': counters,
+        }
+        return metrics
+
+if __name__ == '__main__':
+    import logging
 
-    def configure_callback(self, conf):
-        """Receive configuration block"""
+    logging.basicConfig(level=logging.DEBUG)
+    log = logging.getLogger(__name__)
+    main()
+else:
+    import collectd
+
+    log = collectd
+    amq = AMQMonitor()
+
+    def config_callback(conf):
+        """Collectd config callback."""
         for node in conf.children:
             if node.key == 'Host':
-                self.amq_admin_host = node.values[0]
+                amq.host = node.values[0]
             elif node.key == 'Port':
-                self.amq_admin_port = int(node.values[0])
+                amq.port = int(node.values[0])
             elif node.key == 'Verbose':
-                self.verbose_logging  = bool(node.values[0])
+                amq.verbose_logging = bool(node.values[0])
             else:
-                collectd.warning('activemq_info plugin: Unknown config key: %s.' % node.key)
-        self.log_verbose('Configured with host=%s, port=%s' % (self.amq_admin_host, self.amq_admin_port))
+                log.warning('activemq_info plugin: Unknown config key: %s.'
+                            % node.key)
+        amq.log_verbose('Configured with host={0}, port={1}'.format(
+            (amq.host, amq.port)))
+        collectd.register_read(amq.read_callback)
```

Comment:
This is actually the right thing to do. The read callback shouldn't be registered before the config is done. Otherwise, there's a race condition. In practice, I doubt it would ever matter, but you know about all that academic correctness and all.